# **TAKE HOME TEST PROJECT - NOTEBOOK 1**

# Ecommerce Revenue Leakage & Operational Intelligence



```
Name: Akmal Zuhdy Prasetya
Class: Data Analytics & Business Intelligence
Batch: 19
```

## Libraries

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Datasets

### users.csv

In [ ]:
df_users = pd.read_csv('/content/drive/MyDrive/Bootcamp DABI Batch 19+/program/take home test/raw_data/users.csv')

In [ ]:
df_users.head()

,id,first_name,last_name,email,age,gender,state,street_address,postal_code,city,country,latitude,longitude,traffic_source,created_at
0,457,Timothy,Bush,timothybush@example.net,65,M,Acre,87620 Johnson Hills,69917-400,Rio Branco,Brasil,-9.945568,-67.83561,Search,2022-07-19 13:51:00+00:00
1,6578,Elizabeth,Martinez,elizabethmartinez@example.com,34,F,Acre,1705 Nielsen Land,69917-400,Rio Branco,Brasil,-9.945568,-67.83561,Search,2023-11-08 18:49:00+00:00
2,36280,Christopher,Mendoza,christophermendoza@example.net,13,M,Acre,125 Turner Isle Apt. 264,69917-400,Rio Branco,Brasil,-9.945568,-67.83561,Email,2019-08-24 06:10:00+00:00
3,60193,Jimmy,Conner,jimmyconner@example.com,64,M,Acre,0966 Jose Branch Apt. 008,69917-400,Rio Branco,Brasil,-9.945568,-67.83561,Search,2020-02-15 11:26:00+00:00
4,64231,Natasha,Wilson,natashawilson@example.net,25,F,Acre,20798 Phillip Trail Apt. 392,69917-400,Rio Branco,Brasil,-9.945568,-67.83561,Search,2020-03-13 06:45:00+00:00


In [ ]:
df_users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 15 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   id              100000 non-null  int64  
 1   first_name      100000 non-null  object 
 2   last_name       100000 non-null  object 
 3   email           100000 non-null  object 
 4   age             100000 non-null  int64  
 5   gender          100000 non-null  object 
 6   state           100000 non-null  object 
 7   street_address  100000 non-null  object 
 8   postal_code     100000 non-null  object 
 9   city            99042 non-null   object 
 10  country         100000 non-null  object 
 11  latitude        100000 non-null  float64
 12  longitude       100000 non-null  float64
 13  traffic_source  100000 non-null  object 
 14  created_at      100000 non-null  object 
dtypes: float64(2), int64(2), object(11)
memory usage: 11.4+ MB


### products.csv

In [ ]:
df_products = pd.read_csv('/content/drive/MyDrive/Bootcamp DABI Batch 19+/program/take home test/raw_data/products.csv')

In [ ]:
df_products.head()

,id,cost,category,name,brand,retail_price,department,sku,distribution_center_id
0,13842,2.51875,Accessories,Low Profile Dyed Cotton Twill Cap - Navy W39S55D,MG,6.25,Women,EBD58B8A3F1D72F4206201DA62FB1204,1
1,13928,2.33835,Accessories,Low Profile Dyed Cotton Twill Cap - Putty W39S55D,MG,5.95,Women,2EAC42424D12436BDD6A5B8A88480CC3,1
2,14115,4.87956,Accessories,Enzyme Regular Solid Army Caps-Black W35S45D,MG,10.99,Women,EE364229B2791D1EF9355708EFF0BA34,1
3,14157,4.64877,Accessories,Enzyme Regular Solid Army Caps-Olive W35S45D (...,MG,10.99,Women,00BD13095D06C20B11A2993CA419D16B,1
4,14273,6.50793,Accessories,Washed Canvas Ivy Cap - Black W11S64C,MG,15.99,Women,F531DC20FDE20B7ADF3A73F52B71D0AF,1


In [ ]:
df_products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29120 entries, 0 to 29119
Data columns (total 9 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   id                      29120 non-null  int64  
 1   cost                    29120 non-null  float64
 2   category                29120 non-null  object 
 3   name                    29118 non-null  object 
 4   brand                   29096 non-null  object 
 5   retail_price            29120 non-null  float64
 6   department              29120 non-null  object 
 7   sku                     29120 non-null  object 
 8   distribution_center_id  29120 non-null  int64  
dtypes: float64(2), int64(2), object(5)
memory usage: 2.0+ MB


### orders.csv

In [ ]:
df_orders = pd.read_csv('/content/drive/MyDrive/Bootcamp DABI Batch 19+/program/take home test/raw_data/orders.csv')

In [ ]:
df_orders.head()

,order_id,user_id,status,gender,created_at,returned_at,shipped_at,delivered_at,num_of_item
0,8,5,Cancelled,F,2022-10-20 10:03:00+00:00,NaN,NaN,NaN,3
1,60,44,Cancelled,F,2023-01-20 02:12:00+00:00,NaN,NaN,NaN,1
2,64,46,Cancelled,F,2021-12-06 09:11:00+00:00,NaN,NaN,NaN,1
3,89,65,Cancelled,F,2020-08-13 09:58:00+00:00,NaN,NaN,NaN,1
4,102,76,Cancelled,F,2023-01-17 08:17:00+00:00,NaN,NaN,NaN,2


In [ ]:
df_orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 125226 entries, 0 to 125225
Data columns (total 9 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   order_id      125226 non-null  int64 
 1   user_id       125226 non-null  int64 
 2   status        125226 non-null  object
 3   gender        125226 non-null  object
 4   created_at    125226 non-null  object
 5   returned_at   12530 non-null   object
 6   shipped_at    81461 non-null   object
 7   delivered_at  43884 non-null   object
 8   num_of_item   125226 non-null  int64 
dtypes: int64(3), object(6)
memory usage: 8.6+ MB


### order_items.csv

In [ ]:
df_order_items = pd.read_csv('/content/drive/MyDrive/Bootcamp DABI Batch 19+/program/take home test/raw_data/order_items.csv')

In [ ]:
df_order_items.head()

,id,order_id,user_id,product_id,inventory_item_id,status,created_at,shipped_at,delivered_at,returned_at,sale_price
0,152013,104663,83582,14235,410368,Cancelled,2023-05-07 06:08:40+00:00,NaN,NaN,NaN,0.02
1,40993,28204,22551,14235,110590,Complete,2023-03-14 03:47:21+00:00,2023-03-15 22:57:00+00:00,2023-03-18 01:08:00+00:00,NaN,0.02
2,51224,35223,28215,14235,138236,Complete,2023-12-05 13:25:30+00:00,2023-12-06 01:20:00+00:00,2023-12-10 10:04:00+00:00,NaN,0.02
3,36717,25278,20165,14235,99072,Shipped,2023-12-22 20:48:19+00:00,2023-12-24 16:44:00+00:00,NaN,NaN,0.02
4,131061,90241,71954,14235,353798,Shipped,2022-06-19 16:57:59+00:00,2022-06-19 19:29:00+00:00,NaN,NaN,0.02


In [ ]:
df_order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 181759 entries, 0 to 181758
Data columns (total 11 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   id                 181759 non-null  int64  
 1   order_id           181759 non-null  int64  
 2   user_id            181759 non-null  int64  
 3   product_id         181759 non-null  int64  
 4   inventory_item_id  181759 non-null  int64  
 5   status             181759 non-null  object 
 6   created_at         181759 non-null  object 
 7   shipped_at         118281 non-null  object 
 8   delivered_at       63841 non-null   object 
 9   returned_at        18232 non-null   object 
 10  sale_price         181759 non-null  float64
dtypes: float64(1), int64(5), object(5)
memory usage: 15.3+ MB


### events.csv

In [ ]:
df_events = pd.read_csv('/content/drive/MyDrive/Bootcamp DABI Batch 19+/program/take home test/raw_data/events.csv')

In [ ]:
df_events.head()

,id,user_id,sequence_number,session_id,created_at,ip_address,city,state,postal_code,browser,traffic_source,uri,event_type
0,2198523,NaN,3,83889ed2-2adc-4b9a-af5d-154f6998e778,2021-06-17 17:30:00+00:00,138.143.9.202,São Paulo,São Paulo,02675-031,Chrome,Adwords,/cancel,cancel
1,1773216,NaN,3,7a3fc3f2-e84f-44fe-8876-eff76741f7a3,2020-08-07 08:41:00+00:00,85.114.141.79,Santa Isabel,São Paulo,07500-000,Safari,Adwords,/cancel,cancel
2,2380515,NaN,3,13d9b2fb-eee1-43fd-965c-267b38dd7125,2021-02-15 18:48:00+00:00,169.250.255.132,Mairiporã,São Paulo,07600-000,IE,Adwords,/cancel,cancel
3,2250597,NaN,3,96f1d44e-9621-463c-954c-d8deb7fffe7f,2022-03-30 10:56:00+00:00,137.25.222.160,Cajamar,São Paulo,07750-000,Chrome,Adwords,/cancel,cancel
4,1834446,NaN,3,d09dce10-a7cb-47d3-a9af-44975566fa03,2019-09-05 01:18:00+00:00,161.114.4.174,São Paulo,São Paulo,09581-680,Chrome,Email,/cancel,cancel


In [ ]:
df_events.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2431963 entries, 0 to 2431962
Data columns (total 13 columns):
 #   Column           Dtype  
---  ------           -----  
 0   id               int64  
 1   user_id          float64
 2   sequence_number  int64  
 3   session_id       object 
 4   created_at       object 
 5   ip_address       object 
 6   city             object 
 7   state            object 
 8   postal_code      object 
 9   browser          object 
 10  traffic_source   object 
 11  uri              object 
 12  event_type       object 
dtypes: float64(1), int64(2), object(10)
memory usage: 241.2+ MB


### inventory_items.csv

In [ ]:
df_inventory_items = pd.read_csv('/content/drive/MyDrive/Bootcamp DABI Batch 19+/program/take home test/raw_data/inventory_items.csv')

In [ ]:
df_inventory_items.head()

,id,product_id,created_at,sold_at,cost,product_category,product_name,product_brand,product_retail_price,product_department,product_sku,product_distribution_center_id
0,67971,13844,2022-07-02 07:09:20+00:00,2022-07-24 06:33:20+00:00,2.76804,Accessories,(ONE) 1 Satin Headband,Funny Girl Designs,6.99,Women,2A3E953A5E3D81E67945BCE5519F84C8,7
1,67972,13844,2023-12-20 03:28:00+00:00,NaN,2.76804,Accessories,(ONE) 1 Satin Headband,Funny Girl Designs,6.99,Women,2A3E953A5E3D81E67945BCE5519F84C8,7
2,67973,13844,2023-06-04 02:53:00+00:00,NaN,2.76804,Accessories,(ONE) 1 Satin Headband,Funny Girl Designs,6.99,Women,2A3E953A5E3D81E67945BCE5519F84C8,7
3,72863,13844,2021-10-16 22:58:52+00:00,2021-11-22 02:19:52+00:00,2.76804,Accessories,(ONE) 1 Satin Headband,Funny Girl Designs,6.99,Women,2A3E953A5E3D81E67945BCE5519F84C8,7
4,72864,13844,2021-08-07 16:33:00+00:00,NaN,2.76804,Accessories,(ONE) 1 Satin Headband,Funny Girl Designs,6.99,Women,2A3E953A5E3D81E67945BCE5519F84C8,7


In [ ]:
df_inventory_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 490705 entries, 0 to 490704
Data columns (total 12 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   id                              490705 non-null  int64  
 1   product_id                      490705 non-null  int64  
 2   created_at                      490705 non-null  object 
 3   sold_at                         181759 non-null  object 
 4   cost                            490705 non-null  float64
 5   product_category                490705 non-null  object 
 6   product_name                    490676 non-null  object 
 7   product_brand                   490304 non-null  object 
 8   product_retail_price            490705 non-null  float64
 9   product_department              490705 non-null  object 
 10  product_sku                     490705 non-null  object 
 11  product_distribution_center_id  490705 non-null  int64  
dtypes: float64(2), i

### distribution_centers.csv

In [ ]:
df_distribution_centers = pd.read_csv('/content/drive/MyDrive/Bootcamp DABI Batch 19+/program/take home test/raw_data/distribution_centers.csv')

In [ ]:
df_distribution_centers.head()

,id,name,latitude,longitude
0,1,Memphis TN,35.1174,-89.9711
1,2,Chicago IL,41.8369,-87.6847
2,3,Houston TX,29.7604,-95.3698
3,4,Los Angeles CA,34.0500,-118.2500
4,5,New Orleans LA,29.9500,-90.0667


In [ ]:
df_distribution_centers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   id         10 non-null     int64  
 1   name       10 non-null     object 
 2   latitude   10 non-null     float64
 3   longitude  10 non-null     float64
dtypes: float64(2), int64(1), object(1)
memory usage: 452.0+ bytes


## Tahap 1 — Business Understanding & Problem Statement

### 1.1 Business Background

TheLook merupakan platform ecommerce fashion yang memiliki berbagai aktivitas bisnis seperti:

- customer acquisition,
- user browsing behavior,
- product transactions,
- inventory management,
- order fulfillment,
- hingga product returns.

Meskipun platform memiliki volume transaksi dan aktivitas pengguna yang tinggi, perusahaan berpotensi mengalami revenue leakage akibat:

- rendahnya conversion efficiency,
- performa produk yang kurang optimal,
- keterlambatan fulfillment,
- serta tingginya product return pada kategori tertentu.

Oleh karena itu, diperlukan analisis data end-to-end untuk memahami bagaimana perilaku pelanggan, performa produk, dan efisiensi operasional memengaruhi performa bisnis ecommerce secara keseluruhan.

### 1.2 Problem Statement

Berdasarkan dataset yang tersedia, permasalahan utama yang akan dianalisis dalam proyek ini adalah:

#### **Main Problem Statement**

Bagaimana perilaku pelanggan, performa produk, dan efisiensi operasional memengaruhi revenue leakage dan performa bisnis pada platform ecommerce TheLook?

#### **Supporting Problem Statements**

**1. Customer Acquisition & Behavior**

  - Traffic source mana yang menghasilkan engagement dan conversion terbaik?
  - Bagaimana pola aktivitas dan engagement customer pada platform ecommerce TheLook?
  - Apakah terdapat perbedaan conversion efficiency antar traffic source?

**2. Product & Revenue Performance**

  - Produk dan kategori apa yang memberikan kontribusi revenue tertinggi?
  - Apakah terdapat produk dengan revenue tinggi tetapi profitabilitas rendah?
  - Bagaimana distribusi performa produk berdasarkan category?

**3. Fulfillment & Operational Performance**

  - Bagaimana performa proses fulfillment dan delivery order?
  - Apakah terdapat indikasi operational bottleneck dalam proses pengiriman?
  - Bagaimana hubungan antara fulfillment performance dan order outcome?

**4. Returns & Revenue Leakage**

  - Kategori atau produk apa yang memiliki return rate tertinggi?
  - Seberapa besar potensi revenue leakage akibat product returns?
  - Faktor apa yang berpotensi memengaruhi tingginya return behavior?




### 1.3 Tujuan Analisis

Berdasarkan problem statement yang telah dirumuskan, tujuan utama analisis dalam proyek ini adalah sebagai berikut:

#### **Main Objective**

Menganalisis performa bisnis ecommerce secara end-to-end untuk mengidentifikasi faktor-faktor yang memengaruhi revenue leakage serta menghasilkan insight yang dapat mendukung pengambilan keputusan bisnis.

#### **Specific Objectives**

**1. Customer Acquisition & Analysis**

  - Mengidentifikasi pola customer behavior dan conversion funnel.
  - Menganalisis efektivitas traffic source terhadap engagement dan conversion.
  - Menemukan indikasi conversion inefficiency dalam customer journey.

**2. Product & Revenue Analysis**

  - Menganalisis performa revenue dan profitabilitas produk maupun kategori.
  - Mengidentifikasi produk dengan kontribusi revenue tinggi maupun rendah.
  - Menemukan kategori yang memiliki performa penjualan optimal maupun underperforming.

**3. Fulfillment & Operational Anakysis**

  - Mengevaluasi performa fulfillment dan delivery process.
  - Mengidentifikasi potensi keterlambatan operasional dalam pengiriman order.
  - Menganalisis hubungan antara fulfillment performance dan status order.

**4. Returns & Revenue Leakage Analysis**

  - Mengukur tingkat product return dan dampaknya terhadap bisnis.
  - Mengidentifikasi kategori atau produk dengan return rate tinggi.
  - Menemukan potensi revenue leakage berdasarkan return behavior dan operational performance.




### 1.4 Expected Business Outcome

Hasil analisis diharapkan dapat membantu stakeholder dalam:

- meningkatkan conversion efficiency,
- mengoptimalkan product strategy,
- meningkatkan efisiensi fulfillment,
- mengurangi product return,
- dan meminimalkan revenue leakage pada platform ecommerce.

## Tahap 2 — Data Preprocessing

### 2.1 users.csv

#### **a. Missing Value Check & Handling**

In [ ]:
# missing value check
df_users.isnull().sum()

,0
id,0
first_name,0
last_name,0
email,0
age,0
gender,0
state,0
street_address,0
postal_code,0
city,958


| Check               | Result                    | Interpretation                           |
| ------------------- | ------------------------- | -------------------------------- |
| Missing Values      | Hanya `city`              | Minor handling                   |
| Duplicate Info      | Belum dicek               | Perlu validasi                   |
| Data Types          | `created_at` masih object | Convert datetime                 |
| Numerical Columns   | Sudah sesuai              | Tidak perlu perubahan            |
| Categorical Columns | Sudah usable              | Minor standardization jika perlu |

Missing value hanya pada `city`

Jumlah:

- 958 rows (~0.96%)

Sangat kecil.

Kemudian, karena masih ada:

- `state`,
- `country`,
- `postal_code`,
- `latitude`,
- `longitude`,

maka **tidak perlu drop missing rows tersebut**.

In [ ]:
# missing value handling
# isi dengan "Unknown"
df_users['city'] = df_users['city'].fillna('Unknown')

In [ ]:
# missing value re-check
df_users.isnull().sum()

,0
id,0
first_name,0
last_name,0
email,0
age,0
gender,0
state,0
street_address,0
postal_code,0
city,0


#### **b. Duplicates Check & Handling**

In [ ]:
# general check
df_users.duplicated().sum()

np.int64(0)

In [ ]:
# user's id check
df_users['id'].duplicated().sum()

np.int64(0)

#### **c. Convert `created_at` menjadi datetime**

In [ ]:
# before
df_users['created_at'].head()

,created_at
0,2022-07-19 13:51:00+00:00
1,2023-11-08 18:49:00+00:00
2,2019-08-24 06:10:00+00:00
3,2020-02-15 11:26:00+00:00
4,2020-03-13 06:45:00+00:00


In [ ]:
df_users['created_at'] = pd.to_datetime(df_users['created_at'], format='ISO8601')

In [ ]:
# after
df_users['created_at'].head()

,created_at
0,2022-07-19 13:51:00+00:00
1,2023-11-08 18:49:00+00:00
2,2019-08-24 06:10:00+00:00
3,2020-02-15 11:26:00+00:00
4,2020-03-13 06:45:00+00:00


#### **d. Account Creation Year (Feature Engineering)**

In [ ]:
df_users['account_created_year'] = df_users['created_at'].dt.year

#### **e. Account Creation Month (Feature Engineering)**

In [ ]:
df_users['account_created_month'] = df_users['created_at'].dt.month

#### **f. Kesimpulan Tahap**

In [ ]:
# final preview
df_users.head()

,id,first_name,last_name,email,age,gender,state,street_address,postal_code,city,country,latitude,longitude,traffic_source,created_at,account_created_year,account_created_month
0,457,Timothy,Bush,timothybush@example.net,65,M,Acre,87620 Johnson Hills,69917-400,Rio Branco,Brasil,-9.945568,-67.83561,Search,2022-07-19 13:51:00+00:00,2022,7
1,6578,Elizabeth,Martinez,elizabethmartinez@example.com,34,F,Acre,1705 Nielsen Land,69917-400,Rio Branco,Brasil,-9.945568,-67.83561,Search,2023-11-08 18:49:00+00:00,2023,11
2,36280,Christopher,Mendoza,christophermendoza@example.net,13,M,Acre,125 Turner Isle Apt. 264,69917-400,Rio Branco,Brasil,-9.945568,-67.83561,Email,2019-08-24 06:10:00+00:00,2019,8
3,60193,Jimmy,Conner,jimmyconner@example.com,64,M,Acre,0966 Jose Branch Apt. 008,69917-400,Rio Branco,Brasil,-9.945568,-67.83561,Search,2020-02-15 11:26:00+00:00,2020,2
4,64231,Natasha,Wilson,natashawilson@example.net,25,F,Acre,20798 Phillip Trail Apt. 392,69917-400,Rio Branco,Brasil,-9.945568,-67.83561,Search,2020-03-13 06:45:00+00:00,2020,3


In [ ]:
df_users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 17 columns):
 #   Column                 Non-Null Count   Dtype              
---  ------                 --------------   -----              
 0   id                     100000 non-null  int64              
 1   first_name             100000 non-null  object             
 2   last_name              100000 non-null  object             
 3   email                  100000 non-null  object             
 4   age                    100000 non-null  int64              
 5   gender                 100000 non-null  object             
 6   state                  100000 non-null  object             
 7   street_address         100000 non-null  object             
 8   postal_code            100000 non-null  object             
 9   city                   100000 non-null  object             
 10  country                100000 non-null  object             
 11  latitude               100000 non-null  

- Tahapan preprocessing meliputi pengecekan **missing values**, **duplicate data**, dan **penyesuaian tipe data**.

- Ditemukan missing values pada kolom `city` sebanyak kurang dari 1% dari total data sehingga ditangani menggunakan imputasi kategori “Unknown” untuk mempertahankan konsistensi jumlah data.

- Selain itu, kolom `created_at` diubah ke format datetime untuk mendukung analisis temporal customer acquisition.

- Feature tambahan berupa `account_created_year` dan `account_created_month` juga dibuat untuk mendukung analisis tren pertumbuhan pengguna.

### 2.2 products.csv

#### **a. Missing Value Check & Handling**

In [ ]:
# missing value check
df_products.isnull().sum()

,0
id,0
cost,0
category,0
name,2
brand,24
retail_price,0
department,0
sku,0
distribution_center_id,0


**Main Findings**

| Check             | Result         | Interpretation        |
| ----------------- | -------------- | --------------------- |
| Missing Values    | Sangat sedikit | Minor handling        |
| Duplicate Check   | Belum dicek    | Perlu validasi        |
| Data Types        | Sudah sesuai   | Tidak perlu perubahan |
| Numerical Fields  | Clean          | Siap digunakan        |
| Product Hierarchy | Sudah lengkap  | Sangat bagus          |

**Missing Values Sangat Minimal**
| Column | Missing |
| ------ | ------: |
| `name`   |       2 |
| `brand`  |      24 |

Dari total:

- 29,120 rows

Sangat kecil (<0.1%).

In [ ]:
# missing value handling for "name"
df_products['name'] = df_products['name'].fillna('Unknown Product')

In [ ]:
# missing value handling for "brand"
df_products['brand'] = df_products['brand'].fillna('Unknown Brand')

In [ ]:
# missing value re-check
df_products.isnull().sum()

,0
id,0
cost,0
category,0
name,0
brand,0
retail_price,0
department,0
sku,0
distribution_center_id,0


#### **b. Duplicates Check & Handling**

In [ ]:
# general check
df_products.duplicated().sum()

np.int64(0)

In [ ]:
# product's id check
df_products['id'].duplicated().sum()

np.int64(0)

#### **c. Profit per Product (Feature Engineering)**

In [ ]:
df_products['profit_per_product'] = (
  df_products['retail_price'] - df_products['cost']
)

#### **d. Profit Margin Percentage**

In [ ]:
df_products['profit_margin_pct'] = (
  (df_products['profit_per_product'] / df_products['retail_price']) * 100
)

#### **e. Kesimpulan Tahap**

In [ ]:
df_products.head()

,id,cost,category,name,brand,retail_price,department,sku,distribution_center_id,profit_per_product,profit_margin_pct
0,13842,2.51875,Accessories,Low Profile Dyed Cotton Twill Cap - Navy W39S55D,MG,6.25,Women,EBD58B8A3F1D72F4206201DA62FB1204,1,3.73125,59.7
1,13928,2.33835,Accessories,Low Profile Dyed Cotton Twill Cap - Putty W39S55D,MG,5.95,Women,2EAC42424D12436BDD6A5B8A88480CC3,1,3.61165,60.7
2,14115,4.87956,Accessories,Enzyme Regular Solid Army Caps-Black W35S45D,MG,10.99,Women,EE364229B2791D1EF9355708EFF0BA34,1,6.11044,55.6
3,14157,4.64877,Accessories,Enzyme Regular Solid Army Caps-Olive W35S45D (...,MG,10.99,Women,00BD13095D06C20B11A2993CA419D16B,1,6.34123,57.7
4,14273,6.50793,Accessories,Washed Canvas Ivy Cap - Black W11S64C,MG,15.99,Women,F531DC20FDE20B7ADF3A73F52B71D0AF,1,9.48207,59.3


In [ ]:
df_products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29120 entries, 0 to 29119
Data columns (total 11 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   id                      29120 non-null  int64  
 1   cost                    29120 non-null  float64
 2   category                29120 non-null  object 
 3   name                    29120 non-null  object 
 4   brand                   29120 non-null  object 
 5   retail_price            29120 non-null  float64
 6   department              29120 non-null  object 
 7   sku                     29120 non-null  object 
 8   distribution_center_id  29120 non-null  int64  
 9   profit_per_product      29120 non-null  float64
 10  profit_margin_pct       29120 non-null  float64
dtypes: float64(4), int64(2), object(5)
memory usage: 2.4+ MB


- Pada tabel `products.csv` dilakukan pengecekan missing values, duplicate data, dan validasi tipe data.

- Ditemukan missing values dalam jumlah sangat kecil pada kolom name dan brand sehingga ditangani menggunakan imputasi kategori “Unknown Product” dan “Unknown Brand” untuk menjaga konsistensi data produk.

- Selain itu, dilakukan feature engineering berupa perhitungan `profit_per_product` dan `profit_margin_pct` untuk mendukung analisis profitabilitas produk dan kategori pada tahap exploratory data analysis.

### 2.3 orders.csv

#### **a. Missing Value Check & Handling**

In [ ]:
# missing value check
df_orders.isnull().sum()

,0
order_id,0
user_id,0
status,0
gender,0
created_at,0
returned_at,112696
shipped_at,43765
delivered_at,81342
num_of_item,0


**Main Findings**

| Check              | Result                          | Interpretation      |
| ------------------ | ------------------------------- | ------------------- |
| Missing Values     | Tinggi di timestamp fulfillment | Kemungkinan normal  |
| Data Types         | Timestamp masih object          | Perlu conversion    |
| Duplicate Check    | Belum dicek                     | Wajib validasi      |
| Order Status Logic | Sangat penting                  | Harus dipertahankan |

**Missing Values BUKAN selalu masalah**
 Contoh:

 | Column              | Kemungkinan Makna             |
| ------------------- | ----------------------------- |
| `shipped_at` kosong   | order belum shipped/cancelled |
| `delivered_at` kosong | belum delivered               |
| `returned_at` kosong  | tidak direturn                |

Artinya:

**TIDAK BOLEH asal fill missing values**.

Karena null di sini punya makna bisnis.

Jadi, tidak akan dilakukan filling missing values karena null = status bisnis tertentu. Hal ini penting untuk:

- fulfillment analysis,
- return analysis,
- order lifecycle analysis.

#### **b. Duplicates Check & Handling**

In [ ]:
# general check
df_orders.duplicated().sum()

np.int64(0)

In [ ]:
# orders' id check
df_orders['order_id'].duplicated().sum()

np.int64(0)

#### **c. Convert Timestamp Columns**

In [ ]:
datetime_cols = ['created_at', 'returned_at', 'shipped_at', 'delivered_at']

for col in datetime_cols:
  df_orders[col] = pd.to_datetime(df_orders[col], format='ISO8601')

#### **d. Shipping Duration (Feature Engineering)**

In [ ]:
df_orders['shipping_duration_days'] = (
  df_orders['shipped_at'] - df_orders['created_at']
).dt.days

#### **e. Delivery Duration (Feature Engineering)**

In [ ]:
df_orders['delivery_duration_days'] = (
  df_orders['delivered_at'] - df_orders['shipped_at']
).dt.days

#### **f. Fulfillment Duration (Feature Engineering)**

In [ ]:
df_orders['fulfillment_duration_days'] = (
  df_orders['delivered_at'] - df_orders['created_at']
).dt.days

#### **g. Return Flag (Feature Engineering)**

In [ ]:
df_orders['is_returned'] = (
  df_orders['returned_at'].notnull().astype(int)
)

#### **h. Order Year & Month (Feature Engineering)**

In [ ]:
df_orders['order_year'] = df_orders['created_at'].dt.year
df_orders['order_month'] = df_orders['created_at'].dt.month

#### **i. Kesimpulan Tahap**

In [ ]:
df_orders.head()

,order_id,user_id,status,gender,created_at,returned_at,shipped_at,delivered_at,num_of_item,shipping_duration_days,delivery_duration_days,fulfillment_duration_days,is_returned,order_year,order_month
0,8,5,Cancelled,F,2022-10-20 10:03:00+00:00,NaT,NaT,NaT,3,NaN,NaN,NaN,0,2022,10
1,60,44,Cancelled,F,2023-01-20 02:12:00+00:00,NaT,NaT,NaT,1,NaN,NaN,NaN,0,2023,1
2,64,46,Cancelled,F,2021-12-06 09:11:00+00:00,NaT,NaT,NaT,1,NaN,NaN,NaN,0,2021,12
3,89,65,Cancelled,F,2020-08-13 09:58:00+00:00,NaT,NaT,NaT,1,NaN,NaN,NaN,0,2020,8
4,102,76,Cancelled,F,2023-01-17 08:17:00+00:00,NaT,NaT,NaT,2,NaN,NaN,NaN,0,2023,1


In [ ]:
df_orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 125226 entries, 0 to 125225
Data columns (total 15 columns):
 #   Column                     Non-Null Count   Dtype              
---  ------                     --------------   -----              
 0   order_id                   125226 non-null  int64              
 1   user_id                    125226 non-null  int64              
 2   status                     125226 non-null  object             
 3   gender                     125226 non-null  object             
 4   created_at                 125226 non-null  datetime64[ns, UTC]
 5   returned_at                12530 non-null   datetime64[ns, UTC]
 6   shipped_at                 81461 non-null   datetime64[ns, UTC]
 7   delivered_at               43884 non-null   datetime64[ns, UTC]
 8   num_of_item                125226 non-null  int64              
 9   shipping_duration_days     81461 non-null   float64            
 10  delivery_duration_days     43884 non-null   float64     

Pada tabel `orders.csv` dilakukan pengecekan missing values, duplicate data, dan konversi tipe data.

- Missing values pada kolom `returned_at`, `shipped_at`, dan `delivered_at` tidak dihapus maupun diimputasi karena merepresentasikan kondisi bisnis tertentu, seperti order yang belum dikirim, belum diterima, atau tidak mengalami return.

- Seluruh kolom timestamp dikonversi ke format datetime untuk mendukung analisis temporal dan operasional.

- Selain itu, dilakukan feature engineering berupa `shipping_duration_days`, `delivery_duration_days`, `fulfillment_duration_days`, serta `is_returned` untuk mendukung analisis fulfillment performance dan revenue leakage.

### 2.4 order_items.csv

#### **a. Missing Value Check & Handling**

In [ ]:
# missing value check
df_order_items.isnull().sum()

,0
id,0
order_id,0
user_id,0
product_id,0
inventory_item_id,0
status,0
created_at,0
shipped_at,63478
delivered_at,117918
returned_at,163527


**Main Findings**

| Check           | Result                           | Interpretation   |
| --------------- | -------------------------------- | ---------------- |
| Missing Values  | Tinggi di fulfillment timestamps | Business logic   |
| Data Types      | Timestamp masih object           | Perlu conversion |
| Revenue Data    | `sale_price` tersedia              | Sangat bagus     |


**`order_items.csv` lebih penting daripada `orders.csv`**

Untuk:
- revenue analysis,
- return analysis,
- profitability analysis.

Karena analisis bisnis ecommerce biasanya item-level, bukan order-level.

**Missing Values = Expected Business Logic**

Sama seperti `orders.csv`:

| Column              | Meaning             |
| ------------------- | ------------------- |
| `shipped_at` kosong   | item belum shipped  |
| `delivered_at` kosong | belum delivered     |
| `returned_at` kosong  | item tidak direturn |

Artinya, jangan lakukan imputasi.

#### **b. Duplicates Check & Handling**

In [ ]:
# general check
df_order_items.duplicated().sum()

np.int64(0)

In [ ]:
# order_items' id check
df_order_items['id'].duplicated().sum()

np.int64(0)

#### **c. Convert Timestamp Columns**

In [ ]:
datetime_cols = ['created_at', 'shipped_at', 'delivered_at', 'returned_at']

for col in datetime_cols:
  df_order_items[col] = pd.to_datetime(df_order_items[col], format='ISO8601')

#### **d. Shipping Duration (Feature Engineering)**

In [ ]:
df_order_items['shipping_duration_days'] = (
  df_order_items['shipped_at'] - df_order_items['created_at']
).dt.days

#### **e. Delivery Duration (Feature Engineering)**

In [ ]:
df_order_items['delivery_duration_days'] = (
  df_order_items['delivered_at'] - df_order_items['shipped_at']
).dt.days

#### **f. Fulfillment Duration (Feature Engineering)**

In [ ]:
df_order_items['fulfillment_duration_days'] = (
  df_order_items['delivered_at'] - df_order_items['created_at']
).dt.days

#### **g. Return Indicator (Feature Engineering)**

In [ ]:
df_order_items['is_returned'] = (
  df_order_items['returned_at'].notnull().astype(int)
)

#### **h. Order Year & Month (Feature Engineering)**

In [ ]:
df_order_items['order_year'] = (
  df_order_items['created_at'].dt.year
)

df_order_items['order_month'] = (
  df_order_items['created_at'].dt.month
)

#### **i. Kesimpulan Tahap**

In [ ]:
df_order_items.head()

,id,order_id,user_id,product_id,inventory_item_id,status,created_at,shipped_at,delivered_at,returned_at,sale_price,shipping_duration_days,delivery_duration_days,fulfillment_duration_days,is_returned,order_year,order_month
0,152013,104663,83582,14235,410368,Cancelled,2023-05-07 06:08:40+00:00,NaT,NaT,NaT,0.02,NaN,NaN,NaN,0,2023,5
1,40993,28204,22551,14235,110590,Complete,2023-03-14 03:47:21+00:00,2023-03-15 22:57:00+00:00,2023-03-18 01:08:00+00:00,NaT,0.02,1.0,2.0,3.0,0,2023,3
2,51224,35223,28215,14235,138236,Complete,2023-12-05 13:25:30+00:00,2023-12-06 01:20:00+00:00,2023-12-10 10:04:00+00:00,NaT,0.02,0.0,4.0,4.0,0,2023,12
3,36717,25278,20165,14235,99072,Shipped,2023-12-22 20:48:19+00:00,2023-12-24 16:44:00+00:00,NaT,NaT,0.02,1.0,NaN,NaN,0,2023,12
4,131061,90241,71954,14235,353798,Shipped,2022-06-19 16:57:59+00:00,2022-06-19 19:29:00+00:00,NaT,NaT,0.02,0.0,NaN,NaN,0,2022,6


In [ ]:
df_order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 181759 entries, 0 to 181758
Data columns (total 17 columns):
 #   Column                     Non-Null Count   Dtype              
---  ------                     --------------   -----              
 0   id                         181759 non-null  int64              
 1   order_id                   181759 non-null  int64              
 2   user_id                    181759 non-null  int64              
 3   product_id                 181759 non-null  int64              
 4   inventory_item_id          181759 non-null  int64              
 5   status                     181759 non-null  object             
 6   created_at                 181759 non-null  datetime64[ns, UTC]
 7   shipped_at                 118281 non-null  datetime64[ns, UTC]
 8   delivered_at               63841 non-null   datetime64[ns, UTC]
 9   returned_at                18232 non-null   datetime64[ns, UTC]
 10  sale_price                 181759 non-null  float64     

Pada tabel `order_items.csv` dilakukan pengecekan missing values, duplicate data, serta konversi tipe data timestamp.

- Missing values pada kolom `shipped_at`, `delivered_at`, dan `returned_at` dipertahankan karena merepresentasikan status lifecycle item dalam proses fulfillment dan return.

- Seluruh kolom timestamp dikonversi ke format datetime untuk mendukung analisis temporal dan operasional.

- Selain itu, dilakukan feature engineering berupa `shipping_duration_days`, `delivery_duration_days`, `fulfillment_duration_days`, serta `is_returned` untuk mendukung analisis revenue leakage dan fulfillment performance.

### 2.5 events.csv

#### **a. Missing Value Check & Handling**

In [ ]:
# missing value check
df_events.isnull().sum()

,0
id,0
user_id,1125671
sequence_number,0
session_id,0
created_at,0
ip_address,0
city,23080
state,0
postal_code,0
browser,0


**Main Findings**

| Check           | Result         | Interpretation                 |
| --------------- | -------------- | ------------------------------ |
| Dataset Size    | 2.43 juta rows | Sangat besar                   |
| Missing `user_id` | Sangat tinggi  | Kemungkinan anonymous visitors |
| Missing `city`    | Minor          | Bisa di-handle ringan          |
| Timestamp       | Object         | Perlu conversion               |
| Behavioral Data | Sangat kaya    | Core funnel analysis           |


**NOTE 1: Missing `user_id` kemungkinan besar VALID**

Jumlah missing:

- ±1.1 juta rows (~46%)

Cukup besar, tapi **kemungkinan merepresentasikan anonymous / guest activity**.

Dan dalam ecommerce hal ini normal.

Artinya, **jangan langsung drop rows**.

Karena anonymous traffic tetap penting untuk:

- traffic analysis,
- funnel analysis,
- engagement metrics.

**NOTE 2: `events.csv` TIDAK perlu dibersihkan terlalu agresif**

Karena tujuan kita bukan:

- clickstream engineering kompleks,
- ML session modeling.

Hanya perlu **business intelligence level behavioral analysis**.

#### **b. Duplicates Check & Handling**

In [ ]:
# events' id check
df_events['id'].duplicated().sum()

np.int64(0)

#### **c. Convert `created_at` ke datetime**

In [ ]:
df_events['created_at'] = pd.to_datetime(
  df_events['created_at'],
  format='ISO8601'
)

#### **d. Handle Missing `city`**

In [ ]:
df_events['city'] = df_events['city'].fillna('Unknown')

#### **e. Event Year & Month (Feature Engineering)**

In [ ]:
df_events['event_year'] = (
  df_events['created_at'].dt.year
)

df_events['event_month'] = (
  df_events['created_at'].dt.month
)

#### **f. Event Date (Feature Engineering)**

In [ ]:
df_events['event_date'] = (
  df_events['created_at'].dt.date
)

#### **g. Event Hour (Feature Engineering)**

In [ ]:
df_events['event_hour'] = (
  df_events['created_at'].dt.hour
)

#### **h. Kesimpulan Tahap**

In [ ]:
df_events.head()

,id,user_id,sequence_number,session_id,created_at,ip_address,city,state,postal_code,browser,traffic_source,uri,event_type,event_year,event_month,event_date,event_hour
0,2198523,NaN,3,83889ed2-2adc-4b9a-af5d-154f6998e778,2021-06-17 17:30:00+00:00,138.143.9.202,São Paulo,São Paulo,02675-031,Chrome,Adwords,/cancel,cancel,2021,6,2021-06-17,17
1,1773216,NaN,3,7a3fc3f2-e84f-44fe-8876-eff76741f7a3,2020-08-07 08:41:00+00:00,85.114.141.79,Santa Isabel,São Paulo,07500-000,Safari,Adwords,/cancel,cancel,2020,8,2020-08-07,8
2,2380515,NaN,3,13d9b2fb-eee1-43fd-965c-267b38dd7125,2021-02-15 18:48:00+00:00,169.250.255.132,Mairiporã,São Paulo,07600-000,IE,Adwords,/cancel,cancel,2021,2,2021-02-15,18
3,2250597,NaN,3,96f1d44e-9621-463c-954c-d8deb7fffe7f,2022-03-30 10:56:00+00:00,137.25.222.160,Cajamar,São Paulo,07750-000,Chrome,Adwords,/cancel,cancel,2022,3,2022-03-30,10
4,1834446,NaN,3,d09dce10-a7cb-47d3-a9af-44975566fa03,2019-09-05 01:18:00+00:00,161.114.4.174,São Paulo,São Paulo,09581-680,Chrome,Email,/cancel,cancel,2019,9,2019-09-05,1


In [ ]:
df_events.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2431963 entries, 0 to 2431962
Data columns (total 17 columns):
 #   Column           Dtype              
---  ------           -----              
 0   id               int64              
 1   user_id          float64            
 2   sequence_number  int64              
 3   session_id       object             
 4   created_at       datetime64[ns, UTC]
 5   ip_address       object             
 6   city             object             
 7   state            object             
 8   postal_code      object             
 9   browser          object             
 10  traffic_source   object             
 11  uri              object             
 12  event_type       object             
 13  event_year       int32              
 14  event_month      int32              
 15  event_date       object             
 16  event_hour       int32              
dtypes: datetime64[ns, UTC](1), float64(1), int32(3), int64(2), object(10)
memory usage: 

Pada tabel `events.csv` dilakukan preprocessing dengan fokus pada data behavioral customer.

- Kolom `created_at` dikonversi ke format datetime untuk mendukung analisis temporal dan customer activity trend.

- Missing values pada kolom `city` ditangani menggunakan kategori “Unknown”, sedangkan missing values pada `user_id` dipertahankan karena merepresentasikan aktivitas anonymous user yang masih relevan dalam analisis traffic dan engagement ecommerce.

- Selain itu, dilakukan feature engineering berupa `event_year`, `event_month`, `event_date`, dan `event_hour` untuk mendukung analisis pola aktivitas customer dan traffic behavior.

### 2.6 inventory_items.csv

#### **a. Missing Value Check & Handling**

In [ ]:
# missing value check
df_inventory_items.isnull().sum()

,0
id,0
product_id,0
created_at,0
sold_at,308946
cost,0
product_category,0
product_name,29
product_brand,401
product_retail_price,0
product_department,0


**Main Findings**

| Check                 | Result        | Interpretation   |
| --------------------- | ------------- | ---------------- |
| Missing `sold_at`     | Sangat tinggi | Unsold inventory |
| Missing product_name  | Sangat kecil  | Minor handling   |
| Missing product_brand | Sangat kecil  | Minor handling   |
| Timestamp             | Object        | Perlu conversion |


**NOTE 1: Missing `sold_at` adalah VALID BUSINESS LOGIC**

Jumlah missing:

- ±308 ribu rows

Kemungkinan karena inventory yang belum terjual.

Artinya, jangan dihapus dan jangan diimputasi.

Karena justru penting untuk inventory flow analysis dan stock movement understanding.

**NOTE 2: Tabel ini akan sangat berguna untuk  profitability analysis.**

Karena:

- ada cost,
- retail price,
- sold status,
- product metadata.

Hal ini membuat insight jauh lebih matang.

#### **b. Duplicates Check & Handling**

In [ ]:
# general check
df_inventory_items.duplicated().sum()

np.int64(0)

In [ ]:
# iventory_item's id check
df_inventory_items['id'].duplicated().sum()

np.int64(0)

#### **c. Convert Timestamp Columns**

In [ ]:
datetime_cols = ['created_at', 'sold_at']

for col in datetime_cols:
  df_inventory_items[col] = pd.to_datetime(
  df_inventory_items[col],
  format='ISO8601'
  )

#### **d. Handle Missing Product Name**

In [ ]:
df_inventory_items['product_name'] = (
  df_inventory_items['product_name']
  .fillna('Unknown Product')
)

#### **e. Handle Missing Product Brand**

In [ ]:
df_inventory_items['product_brand'] = (
  df_inventory_items['product_brand']
  .fillna('Unknown Brand')
)

#### **f. Inventory Sold Indicator (Feature Engineering)**

In [ ]:
df_inventory_items['is_sold'] = (
  df_inventory_items['sold_at'].notnull().astype(int)
)

#### **g. Profit per Item (Feature Engineering)**

In [ ]:
df_inventory_items['profit_per_item'] = (
  df_inventory_items['product_retail_price'] -
  df_inventory_items['cost']
)

#### **h. Profit Margin Percentage (Feature Engineering)**

In [ ]:
df_inventory_items['profit_margin_pct'] = (
  (df_inventory_items['profit_per_item'] /
   df_inventory_items['product_retail_price']) * 100
)

#### **i. Inventory Year & Month (Feature Engineering)**

In [ ]:
df_inventory_items['inventory_year'] = (
  df_inventory_items['created_at'].dt.year
)

df_inventory_items['inventory_month'] = (
  df_inventory_items['created_at'].dt.month
)

#### **j. Kesimpulan Tahap**

In [ ]:
df_inventory_items.head()

,id,product_id,created_at,sold_at,cost,product_category,product_name,product_brand,product_retail_price,product_department,product_sku,product_distribution_center_id,is_sold,profit_per_item,profit_margin_pct,inventory_year,inventory_month
0,67971,13844,2022-07-02 07:09:20+00:00,2022-07-24 06:33:20+00:00,2.76804,Accessories,(ONE) 1 Satin Headband,Funny Girl Designs,6.99,Women,2A3E953A5E3D81E67945BCE5519F84C8,7,1,4.22196,60.4,2022,7
1,67972,13844,2023-12-20 03:28:00+00:00,NaT,2.76804,Accessories,(ONE) 1 Satin Headband,Funny Girl Designs,6.99,Women,2A3E953A5E3D81E67945BCE5519F84C8,7,0,4.22196,60.4,2023,12
2,67973,13844,2023-06-04 02:53:00+00:00,NaT,2.76804,Accessories,(ONE) 1 Satin Headband,Funny Girl Designs,6.99,Women,2A3E953A5E3D81E67945BCE5519F84C8,7,0,4.22196,60.4,2023,6
3,72863,13844,2021-10-16 22:58:52+00:00,2021-11-22 02:19:52+00:00,2.76804,Accessories,(ONE) 1 Satin Headband,Funny Girl Designs,6.99,Women,2A3E953A5E3D81E67945BCE5519F84C8,7,1,4.22196,60.4,2021,10
4,72864,13844,2021-08-07 16:33:00+00:00,NaT,2.76804,Accessories,(ONE) 1 Satin Headband,Funny Girl Designs,6.99,Women,2A3E953A5E3D81E67945BCE5519F84C8,7,0,4.22196,60.4,2021,8


In [ ]:
df_inventory_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 490705 entries, 0 to 490704
Data columns (total 17 columns):
 #   Column                          Non-Null Count   Dtype              
---  ------                          --------------   -----              
 0   id                              490705 non-null  int64              
 1   product_id                      490705 non-null  int64              
 2   created_at                      490705 non-null  datetime64[ns, UTC]
 3   sold_at                         181759 non-null  datetime64[ns, UTC]
 4   cost                            490705 non-null  float64            
 5   product_category                490705 non-null  object             
 6   product_name                    490705 non-null  object             
 7   product_brand                   490705 non-null  object             
 8   product_retail_price            490705 non-null  float64            
 9   product_department              490705 non-null  object             
 

Pada tabel `inventory_items.csv` dilakukan preprocessing dengan fokus pada inventory dan profitability analysis.

- Kolom `created_at` dan `sold_at` dikonversi ke format datetime untuk mendukung analisis temporal inventory flow.

-Missing values pada kolom `sold_at` dipertahankan karena merepresentasikan inventory yang belum terjual.

- Missing values dalam jumlah kecil pada kolom `product_name` dan `product_brand` ditangani menggunakan kategori “Unknown Product” dan “Unknown Brand”.

- Selain itu, dilakukan feature engineering berupa `is_sold`, `profit_per_item`, dan `profit_margin_pct` untuk mendukung analisis profitabilitas dan inventory performance.

### 2.7 distribution_centers.csv

#### **a. Missing Value Check & Handling**

In [ ]:
# missing value check
df_distribution_centers.isnull().sum()

,0
id,0
name,0
latitude,0
longitude,0


**Main Findings**

| Check           | Result       | Interpretation       |
| --------------- | ------------ | -------------------- |
| Missing Values  | None         | No action needed     |
| Data Types      | Valid        | No correction needed |
| Dataset Size    | Sangat kecil | Lookup table         |

#### **b. Duplicates Check & Handling**

In [ ]:
# general check
df_distribution_centers.duplicated().sum()

np.int64(0)

In [ ]:
# id check
df_distribution_centers['id'].duplicated().sum()

np.int64(0)

#### **c. Kesimpulan Tahap**

Tabel `distribution_centers.csv` digunakan sebagai lookup table untuk mendukung analisis fulfillment dan operasional distribusi.

Berdasarkan hasil inspeksi data, tidak ditemukan missing values maupun inkonsistensi tipe data sehingga tidak diperlukan proses cleaning tambahan.

Dataset ini siap digunakan untuk proses analisis dan relasi antar tabel.

### 2.8 Analytical Features yang Sudah Dibuat

| Feature                     | Purpose                 |
| --------------------------- | ----------------------- |
| `is_returned`               | return analysis         |
| `fulfillment_duration_days` | operational efficiency  |
| `shipping_duration_days`    | fulfillment KPI         |
| `delivery_duration_days`    | delivery KPI            |
| `profit_per_product`        | profitability analysis  |
| `profit_margin_pct`         | margin analysis         |
| `is_sold`                   | inventory movement      |
| `event_hour`                | customer behavior trend |
| `order_year`/`month`        | temporal analysis       |

Tahapan preprocessing meliputi pengecekan missing values, duplicate data, validasi tipe data, serta feature engineering sederhana untuk mendukung analisis bisnis.

- Pada beberapa tabel transactional seperti `orders` dan `order_items`, missing values pada kolom timestamp tertentu dipertahankan karena merepresentasikan kondisi bisnis seperti order yang belum dikirim atau item yang tidak direturn.

- Selain itu, dilakukan pembuatan beberapa variabel baru seperti `fulfillment_duration_days`, `is_returned`, `profit_margin_pct`, dan `event_hour` untuk mendukung analisis customer behavior, profitability, fulfillment performance, dan revenue leakage.

- Setelah proses preprocessing selesai, seluruh dataset dinyatakan siap digunakan untuk exploratory data analysis dan dashboard development.

#### 2.9 Export Cleaned/Processed Dataset

In [ ]:
df_users.to_csv('users_cleaned.csv', index=False)
df_products.to_csv('products_cleaned.csv', index=False)
df_orders.to_csv('orders_cleaned.csv', index=False)
df_order_items.to_csv('order_items_cleaned.csv', index=False)
df_events.to_csv('events_cleaned.csv', index=False)
df_inventory_items.to_csv('inventory_items_cleaned.csv', index=False)
df_distribution_centers.to_csv('distribution_centers_cleaned.csv', index=False)

## Tahap 3 - EDA (Notebook 2)